lib/endpoints/class-wp-rest-posts-controller.php

```diff
@@ -254,6 +257,36 @@ public function get_item_permissions_check( $request ) {
 	}
 
 	/**
+	 * Can the user access passworded content?
+	 *
+	 * This method determines whether we need to override the regular password
+	 * check in core with a filter.
+	 *
+	 * @param WP_Post $post Post to check against.
+	 * @param WP_REST_Request $request Request data to check.
+	 * @return bool True if the user can access passworded content, false otherwise.
+	 */
+	protected function can_access_password_content( $post, $request ) {
+		if ( empty( $post->post_password ) ) {
+			// No filter required.
+			return false;
+		}
+
+		// Edit context always gets access to passworded posts.
+		if ( $request['context'] === 'edit' ) {
+			return true;
+		}
+
+		// No password, no auth.
+		if ( empty( $request['password'] ) ) {
+			return false;
+		}
+
+		// Double-check the request password.
+		return hash_equals( $post->post_password, $request['password'] );
```

Comment:
Note that this is checked twice: once in the permissions check, and again here. I don't see any huge issues with that, and it makes it a bit cleaner IMO.